arcaflow_plugin_uperf/uperf_plugin.py

```diff
@@ -138,36 +137,65 @@ def process_output(
-    id="uperf_server",
-    name="UPerf Server",
-    description=(
-        "Runs the passive UPerf server to allow benchmarks between the client"
-        " and this server"
-    ),
-    outputs={"success": UPerfServerResults, "error": UPerfServerError},
-)
-def run_uperf_server(
-    params: UPerfServerParams,
-) -> typing.Tuple[str, typing.Union[UPerfServerResults, UPerfServerError]]:
-    # Start the passive server
-    # Note: Uperf calls it 'slave'
-    try:
-        result = subprocess.run(
+class UperfServerStep:
+    exit_event: threading.Event
+
+    def __init__(self):
+        self.exit_event = Event()
+
+    @plugin.signal_handler(
+        id=predefined_schemas.cancel_signal_schema.id,
+        name=predefined_schemas.cancel_signal_schema.display.name,
+        description=predefined_schemas.cancel_signal_schema.display.description,
+        icon=predefined_schemas.cancel_signal_schema.display.icon,
+    )
+    def cancel_step(self, _input: predefined_schemas.cancelInput):
+        # Signal to exit.
+        self.exit_event.set()
+
+    @plugin.step_with_signals(
+        id="uperf_server",
+        name="UPerf Server",
+        description=(
+            "Runs the passive UPerf server to allow benchmarks between the client"
+            " and this server"
+        ),
+        outputs={"success": UPerfServerResults, "error": UPerfServerError},
+        signal_handler_method_names=["cancel_step"],
+        signal_emitters=[],
+        step_object_constructor=lambda: UperfServerStep(),
+    )
+    def run_uperf_server(
+        self,
+        params: UPerfServerParams,
+    ) -> typing.Tuple[str, typing.Union[UPerfServerResults, UPerfServerError]]:
+        # Start the passive server
+        # Note: Uperf calls it 'slave'
+        process = subprocess.Popen(
             ["uperf", "-s"],
             stdout=subprocess.PIPE,
             stderr=subprocess.PIPE,
-            timeout=params.run_duration,
-        )
-        # It should not end itself, so getting here means there was an
-        # error.
-        return "error", UPerfServerError(
-            result.returncode,
-            result.stdout.decode("utf-8") + result.stderr.decode("utf-8"),
         )
-    except subprocess.TimeoutExpired:
-        # Worked as intended. It doesn't end itself, so it finished when it
-        # timed out.
+
+        # Poll for early exit (which is a failure)
+        seconds_ran = 0
+        while seconds_ran < params.run_duration or params.run_duration <= 0:
+            self.exit_event.wait(1.0)
+            # It should not be done yet. If it has exited, it likely failed.
+            exit_return_code = process.poll()
+            if exit_return_code is not None:
+                # There was an error
+                std_out, std_err = process.communicate()
+                return "error", UPerfServerError(
+                    exit_return_code,
+                    std_out.decode("utf-8") + std_err.decode("utf-8"),
+                )
```

Comment:
Instead of using `decode()` here, consider specifying the `text` option or `encoding` option on the `subrocess.Popen()` invocation.

Also, instead of calling `process.communicate()`, consider accessing `stdout` and `stderr` directly, via `process.stdout` and `process.stderr` (the warning about potential deadlocks doesn't apply here, since the target process has already terminated).